### DISCUSSION TITLE
fully dependent types are the only guiding principle for GHC for the foreseeable future(?)

### DISCUSSION THREAD

**Discussion by AntC2** [ghc-proposals/ghc-proposals#462] [General]
(Moving [this comment on a proposal](https://github.com/ghc-proposals/ghc-proposals/pull/452#issuecomment-981282829) to a discussion -- it's more meta than specific design.)

>  Do you have a proposal in mind that is starving?

* Edward Kmett's proposal last year around making `UndecidableInstances` less scary for beginners.
* My proposal last year/year before for Instance guards to avoid overlaps.
* My proposal to not be so eager for error messages to suggest `AllowAmbiguousTypes` -- which leads newbies into a world of pain. Accepted last year, not implemented.
* [This one today](https://github.com/ghc-proposals/ghc-proposals/pull/445) from David F.
* The terrible slow progress on any sort of improvements to the records system. I try to be helpful to Adam, but really he seems to get no support. I note SPJ wrote several joint papers early 2000's, that have gone nowhere.

> Perhaps you have suggestions for improving the process?

We're hostage to the people on the committee who are competent to develop GHC. I feel I've tried all avenues. I'm giving up.

> Unless you are helping guide resources towards GHC (either your own time, your employer's time, your money, your employer's money, etc.) I don't see why you should expect to see developer time directed towards any particular feature you consider to be important.

In a project steered by consensus, it's fair enough for there to be initiatives I'm not interested in. The _quid pro quo_ is that others have to put up with initiatives I'm keen on. I'm getting nothing. What's worse, all this DH stuff (some foundations are already laid) is making my life more complicated. GHC v8 is simply less complicated and already too much of what I don't want and nothing (since ~2015) that I do want.

 > Have you considered getting involved? The development experience is actually pretty good.

I don't even understand the instructions to compile GHC from source. I'm on Windows. I note SPJ has continual messages on the GHC-Devs forum with disasters on Windows. And unlike him, I don't have a brain the size of a planet. I expect my "development experience" would be (expletively) awful.

>  Serious suggestion: have you considered patching Hugs (or even GHC) to support some of the ideas you propose?

Yes I'm tweaking Hugs (as an act of desperation) -- see irregular progress messages on the Hugs forums. Feedback welcome. I have at least got FunDeps+Overlaps working with better coherence than GHC, and such that I can compile all of HList, that Oleg et al had to abandon on Hugs. (But I wouldn't claim it's yet pleasant enough of a user experience.) I should point out I'm not competent to modify a compiler, and prior to that effort had never programmed in C++. I'm saved to the extent it's like the BCPL that I used in 'varsity. Unlike with GHC, I have worked through the instructions; and I can compile WinHugs in Visual Studio. (I deliberately 'funded' this effort by buying a second-hand vintage ~2012 Windows machine.)

**Comment by AntC2** [comment]
> Do you have a proposal in mind that is starving?

Oh, I should wash out my mouth. GHC v9.2.1 brings:

> A new extension  `NoFieldSelectors` hides record field selector functions, so it is possible to define top-level bindings with the same names.

Something I (and others) suggested around 2014, to allow more experimentation wrt field selecting.

**Reply by Ericson2314** [reply]
Yeah `NoFieldSelectors` deserves far more celebration, but it is the small step we really need.

**Comment by goldfirere** [comment]
> In a project steered by consensus, it's fair enough for there to be initiatives I'm not interested in. The quid pro quo is that others have to put up with initiatives I'm keen on. 

I don't quite agree here. In a project steered by consensus, we need to keep the bulk of our users happy. If they are unhappy, they will speak up here and/or leave. So it is in all of our interests to listen to what people say here and take that into account. I don't see this as a "quid pro quo".

This means that the best way to get attention on an issue that lacks that attention is to find others who also think it is a priority. Numbers will speak loudly here.

**Reply by AntC2** [reply]
> we need to keep the bulk of our users happy.

There's a tiny number of people contribute to discussions on proposals. Do you have evidence for whether "the bulk" (who seem to be silent) are (un)happy? Are the silent (non-contributing here) ones broadly happy with the direction of travel? Or maybe they wait until stuff arrives in releases, read the notes and think "nah, not for me". And never use the stuff?

I find it quite intimidating to "speak up here" -- it's just that I'm thick-skinned and have no reputation to lose.

What's the significance of a Proposal that goes all the way through the discussion; gets accepted; then just gathers dust? What is that saying to someone who wrote the Proposal, steered it through the arduous process, kept up the effort because they wanted to make GHC better?

>  If they [our users] are unhappy, they will .../or leave.

I don't suppose many of those who leave would go to the bother of saying so or explaining why. They'll just fade away. Perhaps GHC HQ wants those not drinking the Dependent Haskell kool-aid to leave?

I can't see any motivation for DH, but I would be prepared to rub along with it if I was getting other features I am motivated about. But I think DH will fully occupy GHC for the "foreseeable future"; it's an entirely reasonable observation that it's "the only guiding principle for GHC".

**Reply by goldfirere** [reply]
> Do you have evidence for whether "the bulk" (who seem to be silent) are (un)happy?

No. One outcome I am hoping for with the Haskell Foundation is that we have more reliable ways of learning what our users like and dislike about the language. But we have to go with what we have: only those users who do speak up will be counted. That's far from ideal, but I don't know of another way to do open-source development.

> What is that saying to someone who wrote the Proposal, steered it through the arduous process, kept up the effort because they wanted to make GHC better?

A good question. One possibility is that we accept only those proposals that have a ready implementor. We have decided not to do that -- to keep proposal acceptance and implementation separate. Maybe it's the wrong choice, because of the negative feeling of authors of accepted proposals.

I still say: if you want an accepted proposal implemented, go find others who agree with you. Get them to post on the proposal or ticket that they are eager for it. That will create change. As it is, an accepted, unimplemented proposal (that stays that way) suggests that it describes a feature that the committee thinks is "nice to have", but the implementors think not worth the development cost. (There is overlap, but not coincidence, between the committee and the set of implementors.) To change that fact, get more people to say how badly they want the feature.

Is it intimidating to speak up here? Perhaps. But people really can get action just with thumbs-ups! (and then maybe a follow up post to cause a notification)

> I don't suppose many of those who leave would go to the bother of saying so or explaining why. They'll just fade away.

This is true. Again, I don't really know how to counteract this fact. You seem to be suggesting we mix DH-related progress in with other ideas. But maybe that decision would cause more people to fade away. We just don't know.

**Reply by Ericson2314** [reply]
@AntC2 so proposals being accepted wasn't supposed to go indicate someone will dutifully implement them. Your https://github.com/ghc-proposals/ghc-proposals/blob/master/proposals/0232-AmbiguousType-pragma.rst#implementation-plan punts on this other than to say you don't think you can do it, and so it is likewise undone. I also have an accepted proposal #285 which I didn't get around to implementing, but it's mere existence helps ground #448 which I hope will finally put this stuff to rest, and so I am pleased with the outcome.

**Reply by adamgundry** [reply]
My impression is that there is a bit of a problem here, whereby accepted proposals can languish without an implementer. This is somewhat difficult to solve, because we can't magically conjure up (volunteer) GHC developers to work on arbitrary features. However, it risks a gulf between proposals by "insiders" (active GHC devs) which get implemented promptly once accepted, and those by "outsiders" which may be accepted but then sit around making no progress. The latter are particularly problematic if later developments invalidate parts of the proposal, so when someone subsequently comes to implement it, they may need to re-design some aspects.

I'm not sure if it is even easy to get a list of proposals that have been accepted but not implemented? In principle the pull requests against this repo are labelled, but those labels aren't always up to date and it is restricted who can edit them. The GHC issue tracker contains issues corresponding to some but not all accepted proposals.

It's not obvious to me how to solve this, but I do think there are some things we could potentially do to make the situation better:
 * Require that every accepted proposal must have an associated GHC ticket (i.e. the field should be filled in before merging the PR, and someone should create tickets for all the existing accepted proposals). Then we can use the GHC issue tracker to list accepted-but-not-implemented proposals, and GHC devs looking for things to do will encounter them.
 * Once a proposal has been accepted, make an active effort to find implementers. I don't think having an implementer lined up should be a precondition for acceptance, but where the author isn't in a position to implement themselves, perhaps the committee could help e.g. by advertising the project. Starting an implementation before final acceptance would be ideal, because often there are corner cases of the design that are only really exposed during the implementation work, and those may or may not be reflected back in the proposal if the proposal has long since been accepted.
 * (Perhaps) if volunteer efforts are not forthcoming, GHC HQ might reasonably dedicate paid engineering time to implementing accepted proposals.

**Reply by nomeata** [reply]
I am not sure we actually do have a problem (besides maybe one of wrong expectations). Even without the GHC committee to aid in deciding what would be a good change, only what has implementors gets implemented. And it's natural in a mostly volunteer driven project it's natural that only those things that find volunteers have a chance at implementing. Whether it's a small project where the single project developer comments with “sounds good, PR welcome”, or a whole committee says “this change seems desirable” doesn't really change things, does it?

So as long as it's clear to all involved that implementation work is still on those who care enough about a change, it seems it seems the status quo is fine, is it?

**Reply by tomjaguarpaw** [reply]
I'm not involved in GHC development but as an interested outsider I agree with Joachim. If "accepted GHC proposal" meant "someone in the GHC team is *required* to implement your proposal" then I think we'd find ourselves with many fewer accepted proposals.

**Reply by AntC2** [reply]
Thanks all for the feedback. You're confirming for me "fully dependent types are the only guiding principle for GHC for the foreseeable future"; simply because no developer of GHC is interested in/prepared to put their inevitably limited time into anything else. I'm feeling justified in abandoning GHC v9 onwards.

I think the proposals process should say on the tin: If you don't have a resource/competence to develop this, don't waste everybody's time proposing it.

> @goldfirere  One possibility is that we accept only those proposals that have a ready implementor. We have decided not to do that -- to keep proposal acceptance and implementation separate. Maybe it's the wrong choice, because of the negative feeling of authors of accepted proposals.

Yes wrong choice: long before acceptance, indeed long before an author writes up a proposal, if no implementor steps forward, just shut down the proposal. No point in taking it beyond the stage of a 'bright idea'. "Negative feelings" of not only the author but every reader/commenter who puts effort into trying to understand it and polish the proposal.

 And perhaps someone could discourage SPJ from remarks like ["I think there is material for a good GHC Proposal here."](https://gitlab.haskell.org/ghc/ghc/-/issues/12203#note_340847). (I'm feeling justified in not pursuing that one.)

**Reply by tomjaguarpaw** [reply]
The proposals template has an ["Implementation Plan" section](https://github.com/ghc-proposals/ghc-proposals/blob/master/proposals/0000-template.md#implementation-plan) but it is labelled "optional".  Perhaps it should be mandatory?

**Reply by nomeata** [reply]
> simply because no developer of GHC is interested in/prepared to put their inevitably limited time into anything else.

I hope this was not quite meant as it can sound, given the amazing, but maybe less visible, work that GHC developers put into performance, cross-compilation, code generation, core-to-core optimization, HLS integration and plain bug fixing (and I am sure to have forgotten stuff)

**Reply by adamgundry** [reply]
I thought a key goal of the proposal process was to solicit input from Haskellers who are not compiler developers. We should be trying to include them, not exclude them!

> I am not sure we actually do have a problem (besides maybe one of wrong expectations). Even without the GHC committee to aid in deciding what would be a good change, only what has implementors gets implemented. And it's natural in a mostly volunteer driven project it's natural that only those things that find volunteers have a chance at implementing. Whether it's a small project where the single project developer comments with “sounds good, PR welcome”, or a whole committee says “this change seems desirable” doesn't really change things, does it?
> 
> So as long as it's clear to all involved that implementation work is still on those who care enough about a change, it seems it seems the status quo is fine, is it?

@nomeata I'm going to politely disagree here. The point is that proposals themselves don't find volunteers, people do. :wink: And the question is who should bear that responsibility. Even if the committee or GHC HQ can't guarantee to find volunteers, they can potentially help, and may be in a better position to identify and motivate existing or new contributors than the proposer.

At the moment the status quo (as I understand it) is that once a proposal is accepted, the committee considers that its job is done, and it falls on the proposer to:
 * open a GHC ticket (and make sure it gets added to the proposal),
 * implement it or find someone else to do so, and
 * report back when the proposal is implemented.

My concrete suggestion is that it becomes part of the process to open a GHC issue, and link to it in the proposal, once the proposal is accepted but before the PR is merged. This will bring it to the attention of the GHC devs (at least those who follow the GHC issue tracker), and makes it clear what the status is (e.g. is the ticket unassigned, in progress, or resolved). That doesn't fully solve the problem of finding volunteers to implement, but it would be a start.

**Reply by nomeata** [reply]
Opening tickets as part of the process is a useful refinement, I don't disagree here :-). But open issues in GitLab may change visibility, but not the fundamental “it takes someone who cares about it”

**Reply by goldfirere** [reply]
I would support opening an issue in GitLab to be part of our "acceptance" process.

Another way to increase discoverability: if we are confident, say, that all accepted proposals are linked to their GHC tickets and that those tickets are up to date (tickets do tend to stay up to date, actually), then we can imagine an "unimplemented" tag in this repo to help potential contributors find proposals that need implementing. (If they find a proposal labeled "unimplemented" and then follow it to the related ticket, they can quickly discover whether the label is out of date.)

Again this doesn't solve the underlying problem... but it maybe makes it a little better. Incremental improvement is superior to no improvement.

**Reply by adamgundry** [reply]
It is possible to list [proposals that are labelled accepted but not implemented](https://github.com/ghc-proposals/ghc-proposals/pulls?q=is%3Apr+label%3Aaccepted+-label%3Aimplemented), although this list is not completely up to date.

**Reply by AntC2** [reply]
> “it takes someone who cares about it”

Is there any (potential) GHC developer who "cares about" anything but Dependent Haskell?

> the underlying problem

is it that there simply aren't enough people with time/skill to develop GHC? Or there are plenty of adequately skilled volunteers but they don't know how to dip their toes in?

Amongst that list of accepted/not implemented (thanks @adamgundry ) there looks to be plenty of non-DH items. Also some that are more 'polishing' existing functionality than major chunks of new features. Could that be a path for dipping toes in? (I guess they'd still need oversight, which takes it right back to "the underlying problem". And no way to get a published paper from the polishing.)

**Reply by utdemir** [reply]
> Is there any (potential) GHC developer who "cares about" anything but Dependent Haskell?

> You're confirming for me "fully dependent types are the only guiding principle for GHC for the foreseeable future"; simply because no developer of GHC is interested in/prepared to put their inevitably limited time into anything else.

@AntC2 ,

I'm not a GHC developer but a mere Haskell user, but I couldn't stop myself pitching in. Just as an example if you look at the [last](https://downloads.haskell.org/ghc/9.0.1/docs/html/users_guide/9.0.1-notes.html) [two](https://ghc.gitlab.haskell.org/ghc/doc/users_guide/9.2.1-notes.html) major GHC releases; you could see that there are many great features, and almost none relevant to Dependent Haskell. And on top of this, check out the [GHC 9.4 milestone](https://gitlab.haskell.org/ghc/ghc/-/milestones/370).

If you look at any other (healthy) language ecosystem [[1]] [[2]], there are both many proposals under discussion for a long time, and also many unimplemented accepted proposals. I'm not saying that things can not be improved, but it is not accurate to think that "there is no development occurring because of this specific feature".

I can see how it can be frustrating to have a proposal you care about not moving forward. I would also love for every proposal to get the attention they deserve, and as an outside observer, people in this thread is actively discussing how to make things better. But you still seem to singling-out the dependent types story like the entire GHC development team (as such a thing exists) is only focusing on it, which is evidently not the case looking at the GHC commits and other proposals that are moving forward.

tl;dr: I do not know why you are singling out Dependent Haskell every single time, but it is almost disrespectful both to the users like me and the developers to say that GHC is not moving forward. So let's not dismiss all the hard work of the GHC contributors working on just because we do not like a specific feature some of them might/will work on at some point in future.

[1]: https://www.python.org/dev/peps/#id9
[2]: https://github.com/rust-lang/rfcs/pulls
